pkg/wasm/cache.go

```diff
@@ -104,25 +106,41 @@ func NewLocalFileCache(dir string, purgeInterval, moduleExpiry time.Duration, in
 	return cache
 }
 
+func needToUseCache(pullPolicy extensions.PullPolicy, u *url.URL) bool {
+	if u.Scheme == "oci" {
+		switch pullPolicy {
+		case extensions.PullPolicy_Always:
+			return false
+		case extensions.PullPolicy_IfNotPresent:
+			return true
+		default:
+			return !strings.HasSuffix(u.Path, ":latest")
+		}
+	}
+	return true
+}
+
 // Get returns path the local Wasm module file.
-func (c *LocalFileCache) Get(downloadURL, checksum string, timeout time.Duration, pullSecret []byte) (string, error) {
+func (c *LocalFileCache) Get(downloadURL, checksum string, timeout time.Duration, pullSecret []byte, pullPolicy extensions.PullPolicy) (string, error) {
 	// Construct Wasm cache key with downloading URL and provided checksum of the module.
 	key := cacheKey{
 		downloadURL: downloadURL,
 		checksum:    checksum,
 	}
 
-	// First check if the cache entry is already downloaded.
-	if modulePath := c.getEntry(key); modulePath != "" {
-		return modulePath, nil
-	}
-
-	// If not, fetch images.
 	u, err := url.Parse(downloadURL)
 	if err != nil {
 		return "", fmt.Errorf("fail to parse Wasm module fetch url: %s", downloadURL)
 	}
 
+	// First check if the cache entry is already downloaded and policy does not require to pull always.
```

Comment:
What does it mean to 'Always' pull? In Kubernetes it is clear - when we start up a pod. What does it mean for WASM though? Each time the xDS changes?

Reply:
The API docs say simply: `'Always' is the default, mirroring K8s behaviour.` This _could_ be interpreted as lifetime of proxy container. That seems like it would essentially be the same as `IfNotPresent` then, however. 

The comment on `Always` suggests: `We will always pull the latest version of an image when applying this plugin.` Which I think means with each xDS push.

Thoughts?

Reply:
Pulling on every xDS push seems not great - we push a lot

Reply:
Yes. Current implementation updates for each time of ECDS change for 'Always'.

If there is any desirable interpretation, please let me know that.

Reply:
> Pulling on every xDS push seems not great - we push a lot

Use of `Always` considered harmful ??

Reply:
Even with `Always`, if we check sha256, the number of xDS would not be problematic (of course, when the users provide sha256).

Please note that current implementation re-pulls the oci image or binary (via http) for each xDS change when the user does not provide sha256.

Reply:
We can use sha256 specified in `WasmPlugin` resource without pulling the config of images.
In normal case, I think it would be enough. 
Do you think that it is needed to care about mis-configuration as well?

Reply:
I am sorry. I realized that I am confused a bit.

As I mentioned above, currently if a user does not provide sha256 in WasmPlugin, it works like `Always` (pulling for each xDS changes). 

At this point, do you think current `ImagePullPolicy` is necessary? 

Rather, I guess that it would be better to remove `ImagePullPolicy` or add new property like `ImageExpiration` (or `ImageMinLifetime` whatever).

WDYT?

Reply:
One idea:

The comment for `Always` is `We will always pull the latest version of an image when applying this plugin`, 
so how about defining `Always` as `each time when WasmPlugin resource is changed`?

In implementation-wise, I think we can use "Namespace+Name(of WasmPlugin)+ResourceVersion" as a part of key in the local image cache.

Reply:
I like the idea. We should check if ResourceVersion changes when annotation changes, since one way users may try to trigger a new pull is changing an annotation? It might, I just forget the semantics of the field

Reply:
Confirmed. The change of annotations causes ResourceVersion change. 

BTW, just for checking, Istiod reacts the change of annotation as well?

Reply:
As discussed here, I have changed this part to pull the image for each time that `ResourceVersion` is changed.